src/optimagic/differentiation/derivatives.py

```diff
@@ -1,53 +1,110 @@
 import itertools
 import re
+from dataclasses import dataclass
 from itertools import product
-from typing import NamedTuple
+from typing import Any, Callable, Literal, NamedTuple, cast
 
 import numpy as np
 import pandas as pd
+from numpy.typing import NDArray
 from pybaum import tree_flatten, tree_unflatten
 from pybaum import tree_just_flatten as tree_leaves
 
-from optimagic import batch_evaluators
+from optimagic import batch_evaluators, deprecations
 from optimagic.config import DEFAULT_N_CORES
-from optimagic.deprecations import replace_and_warn_about_deprecated_bounds
+from optimagic.deprecations import (
+    replace_and_warn_about_deprecated_base_steps,
+    replace_and_warn_about_deprecated_bounds,
+)
 from optimagic.differentiation import finite_differences
 from optimagic.differentiation.generate_steps import generate_steps
 from optimagic.differentiation.richardson_extrapolation import richardson_extrapolation
 from optimagic.parameters.block_trees import hessian_to_block_tree, matrix_to_block_tree
 from optimagic.parameters.bounds import Bounds, get_internal_bounds, pre_process_bounds
 from optimagic.parameters.tree_registry import get_registry
+from optimagic.typing import PyTree
+
+
+@dataclass(frozen=True)
+class NumdiffResult:
+    """Result of a numerical differentiation.
+
+    The following relationship holds for vector-valued functions with vector-valued
+    parameters:
+
+    First Derivative:
+    -----------------
+
+    - f: R -> R leads to shape (1,), usually called derivative
+    - f: R^m -> R leads to shape (m, ), usually called Gradient
+    - f: R -> R^n leads to shape (n, 1), usually called Jacobian
+    - f: R^m -> R^n leads to shape (n, m), usually called Jacobian
+
+    Second Derivative:
+    ------------------
+
+    - f: R -> R leads to shape (1,), usually called second derivative
+    - f: R^m -> R leads to shape (m, m), usually called Hessian
+    - f: R -> R^n leads to shape (n,), usually called Hessian
+    - f: R^m -> R^n leads to shape (n, m, m), usually called Hessian tensor
+
+    Attributes:
+        derivative: The estimated derivative at the parameters. The structure of the
+            derivative depends on the input parameters and the output of the function.
+        func_value: The value of the function at the parameters.
+
+    """
+
+    derivative: PyTree
+    func_value: PyTree | None = None
+    # deprecated
+    _func_evals: pd.DataFrame | dict[str, pd.DataFrame | None] | None = None
+    _derivative_candidates: pd.DataFrame | None = None
+
+    @property
+    def func_evals(self) -> pd.DataFrame | dict[str, pd.DataFrame | None] | None:
+        deprecations.throw_numdiff_result_func_evals_future_warning()
+        return self._func_evals
+
+    @property
+    def derivative_candidates(self) -> pd.DataFrame | None:
+        deprecations.throw_numdiff_result_derivative_candidates_future_warning()
+        return self._derivative_candidates
+
+    def __getitem__(self, key: str) -> Any:
```

Comment:
I think we should deprecate the dict access right away. It is only there so less old code is broken.